# Connect to google account

In [1]:
# 구글 드라이브 연결
from google.colab import drive
drive.mount("/content/data")

Mounted at /content/data


In [2]:
# 모듈 Import
import numpy as np # 행렬 계산에 사용하는 모듈
import pandas as pd # 데이터 처리와 분석을 위한 모듈
import matplotlib.pyplot as plt # 데이터 시각화를 위한 모듈. 2D, 3D 그릴 때 사용
import seaborn as sns # 데이터 시각화를 위한 모듈. 두 데이터의 관계를 볼때 사용
from sklearn.metrics.pairwise import cosine_similarity

In [3]:
# 딕셔너리를 간편하게 사용하는 것을 도와준다. 모델 class 처럼 사용 가능
# dot을 이용해 객체를 불러 사용. JSON 다룰때 유용.
import easydict
args = easydict.EasyDict()

# path
args.default_path = "/content/data/MyDrive/Playdata/Competitions/ML/Dacon/" # 메인 경로
args.apply_train_path = args.default_path + "apply_train.csv" # train 데이터 경로
args.company_path = args.default_path + "company.csv"
args.recruitment_path = args.default_path + "recruitment.csv"
args.resume_certificate_path = args.default_path + "resume_certificate.csv"
args.resume_education_path = args.default_path + "resume_education.csv"
args.resume_language_path = args.default_path + "resume_language.csv"
args.resume_path = args.default_path + "resume.csv"

args.default_submission_path = args.default_path + "sample_submission.csv" # 예측결과(제출파일) 경로

# 데이터 분석을 위한 변수들
# # 난수 생성 제어 => 같은 코드를 실행해도 동일한 결과를 얻기 위해서 설정
# 데이터 분할 및 모델 초기화 때 유용.
args.random_state = 42
args.results = [] # 결과 저장 리스트

In [4]:
args.submission_path = args.default_path + "result/submission_baseLine_2.csv" # 결과 저장 파일
args.save_results = args.default_path+"result/model_results_baseLine_2.json" # 결과 저장 json

# File Load

In [5]:
apply_train_df = pd.read_csv(args.apply_train_path) # apply_train -> DataFrame화
company_df = pd.read_csv(args.company_path) # company -> DataFrame화
recruitment_df = pd.read_csv(args.recruitment_path) # recruitment -> DataFrame화
resume_certificate_df = pd.read_csv(args.resume_certificate_path) # resume_certificate -> DataFreame화
resume_education_df = pd.read_csv(args.resume_education_path) # resume_education -> DataFrame화
resume_language_df = pd.read_csv(args.resume_language_path) # resume_language -> DataFrame화
resume_df = pd.read_csv(args.resume_path) # resume -> DataFrame화


In [6]:
# 예측할 파일 정보
pd.read_csv(args.default_submission_path).shape

(42410, 2)

In [7]:
apply_train_df

resume_seq recruitment_seq
0         U05833          R03838
1         U06456          R02144
2         U07807          R01877
3         U04842          R02463
4         U08336          R00112
...          ...             ...
57941     U02270          R03430
57942     U02640          R04987
57943     U08238          R01342
57944     U01296          R06363
57945     U05748          R03090

[57946 rows x 2 columns]

- 57946개의 채용 지원 존재

# User-Item Matrix / Similarity / Score

In [8]:
# 사용자-아이템 행렬 생성: 구직자가 해당 채용 공고에 지원했으면 1, 아니면 0으로 설정
user_item_matrix = apply_train_df.groupby(['resume_seq', 'recruitment_seq']).size().unstack(fill_value=0)

In [9]:
user_item_matrix

recruitment_seq  R00001  R00002  R00003  R00004  R00005  R00006  R00007  \
resume_seq                                                                
U00001                0       0       0       0       0       0       0   
U00002                0       0       0       0       0       0       0   
U00003                0       0       0       0       0       0       0   
U00004                0       0       0       1       0       0       0   
U00005                0       0       0       0       0       0       0   
...                 ...     ...     ...     ...     ...     ...     ...   
U08478                0       0       0       0       0       0       0   
U08479                0       0       0       0       0       0       0   
U08480                0       0       0       0       0       0       0   
U08481                0       0       0       0       0       0       0   
U08482                0       0       0       0       0       0       0   

recruitment_seq  R00008  R00009  R00010  ...  R06686  R06687  R06688  R06689  \
resume_seq                               ...                                   
U00001                0       0       0  ...       0       0       0       0   
U00002                0       0       0  ...       0       0       0       0   
U00003                0       0       0  ...       0       0       0       0   
U00004                0       0       0  ...       0       0       0       0   
U00005                0       0       0  ...       0       0       0       0   
...                 ...     ...     ...  ...     ...     ...     ...     ...   
U08478                0       0       0  ...       0       0       0       0   
U08479                0       0       0  ...       0       0       0       0   
U08480                0       0       0  ...       0       0       0       0   
U08481                0       0       0  ...       0       0       0       0   
U08482                0       0       0  ...       0       0       0       0   

recruitment_seq  R06690  R06691  R06692  R06693  R06694  R06695  
resume_seq                                                       
U00001                0       0       0       0       0       0  
U00002                0       0       0       0       0       0  
U00003                0       0       0       0       0       0  
U00004                0       0       0       0       0       0  
U00005                0       0       0       0       0       0  
...                 ...     ...     ...     ...     ...     ...  
U08478                0       0       0       0       0       0  
U08479                0       0       0       0       0       0  
U08480                0       0       0       0       0       0  
U08481                0       0       0       0       0       0  
U08482                0       0       0       0       0       0  

[8482 rows x 6695 columns]

In [10]:
# 채용 횟수가 2번 이상이면 1로 변경
user_item_matrix[user_item_matrix > 1] = 1

In [11]:
# 사용자 간의 유사성 계산
user_similarity = cosine_similarity(user_item_matrix)

In [12]:
user_similarity

array([[1., 0., 0., ..., 0., 0., 0.],
       [0., 1., 0., ..., 0., 0., 0.],
       [0., 0., 1., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 1., 0., 0.],
       [0., 0., 0., ..., 0., 1., 0.],
       [0., 0., 0., ..., 0., 0., 1.]])

In [13]:
# 추천 점수 계산
user_predicted_scores = user_similarity.dot(user_item_matrix) / np.array([np.abs(user_similarity).sum(axis=1)]).T

In [14]:
user_predicted_scores

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])

In [15]:
user_predicted_scores.shape

(8482, 6695)

# Prediction

In [ ]:
# 이미 지원한 채용 공고 제외하고 추천
recommendations = []
for idx, user in enumerate(user_item_matrix.index):
    # 해당 사용자가 지원한 채용 공고
    applied_jobs = set(user_item_matrix.loc[user][user_item_matrix.loc[user] == 1].index)

    # 해당 사용자의 추천 점수 (높은 점수부터 정렬)
    sorted_job_indices = user_predicted_scores[idx].argsort()[::-1]
    recommended_jobs = [job for job in user_item_matrix.columns[sorted_job_indices] if job not in applied_jobs][:5]

    for job in recommended_jobs:
        recommendations.append([user, job])

# Submission

In [ ]:
# sample_submission.csv 형태로 DataFrame 생성
top_recommendations = pd.DataFrame(recommendations, columns=['resume_seq', 'recruitment_seq'])

top_recommendations.to_csv(args.submission_path, index=False)

# 데이터 전처리 (train data, test data)

## EDA (train data)

## 결측치 제거

## 신규컬럼 생성

## Encoding

## Scaling

# 앙상블 Modeling (train_f, train_t)
- shape, 결측치 확인 필수

## 데이터 분리 (train data)

## Experiment

- 필요한 라이브러리 import

### Cross Validation

### Model

### HPO

## Best Model 해석

### Cross Validation

### Best Model 생성 및 학습

### Confusion Matrix

### HeatMap by Confusion Matrix

# 예측

# Result 저장 (Submission)
- Dictionary List로 sort해서 best model select